File: q3100/q3100.java
class Solution {
    public int maxBottlesDrunk(int numBottles, int numExchange) {
        int res = 0;
        int currFull = numBottles;
        int currEmpty = 0;
        while (currFull != 0 || currEmpty >= numExchange) {
            // greedy - try to exchange whenever possible
            if (currEmpty >= numExchange) {
                currFull++;
                currEmpty -= numExchange;
                numExchange++;
            }

            // drink all water
            currEmpty += currFull;
            res+=currFull;
            currFull = 0;
        }
        return res;
    }
}

File: q3100/q3100.py
class Solution:
    def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
        total = 0 
        currFull = numBottles
        currEmpty = 0

        while currFull or currEmpty >= numExchange:
            if currEmpty >= numExchange: 
                currEmpty -= numExchange
                currFull += 1
                numExchange += 1

            currEmpty += currFull
            total += currFull
            currFull = 0
        
        return total
